pd/src/components/shielded_pool.rs

```diff
@@ -321,3 +321,54 @@ impl ShieldedPool {
         }
     }
 }
+
+/// Extension trait providing read/write access to shielded pool data.
+///
+/// TODO: should this be split into Read and Write traits?
+#[async_trait]
+pub trait ShieldedPoolStore: WriteOverlayExt {
+    async fn token_supply(&self, asset_id: &asset::Id) -> Result<Option<u64>> {
+        self.get_proto(format!("shielded_pool/assets/{}/token_supply", asset_id).into())
+            .await
+    }
+
+    #[instrument(skip(self))]
+    async fn update_token_supply(&self, asset_id: &asset::Id, change: i64) -> Result<()> {
+        let key = format!("shielded_pool/assets/{}/token_supply", asset_id).into();
+        let current_supply = match self.get_proto(key).await {
+            Ok(Some(value)) => value,
+            Ok(None) => 0u64,
+            // We want to handle the MissingRootError specially here, so that we can
+            // use 0 as a default value if the tree is empty, without ignoring all errors.
+            Err(e) if e.downcast_ref::<jmt::MissingRootError>().is_some() => 0u64,
+            Err(e) => return Err(e),
+        };
```

Comment:
cc @aubrika ; IIRC we had a discussion about using downcasting inside of the `WriteOverlay` in the JMT crate.

 I think it'd be much cleaner if the `WriteOverlay` did the downcasting itself, and always replaced `Err(MissingRootError)` with `Ok(None)`, so we could get this behavior everywhere.